skbio/tree/tests/test_nj.py

```diff
@@ -191,6 +221,201 @@ def test_pair_members_to_new_node_zero_branch_length(self):
         self.assertEqual(
             _pair_members_to_new_node(dm, 'a', 'b', False), (-16, 20))
 
+    def test_nni_dm1(self):
+        self.assertEqual(nj(self.dm1, result_constructor=str),
+                         self.expected1_str)
+        actual_TreeNode = nni(self.pre1_nni_TreeNode, self.dm1, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post1_nni_TreeNode), 0.0, places=10)
+
+    def test_nni_dm2(self):
+        # Resulting tree topology is equivalent to result from nj, however,
+        # resulting edge lengths are almost equal to 2 places.
+        actual_TreeNode = nni(self.pre2_nni_TreeNode, self.dm2, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post2_nni_TreeNode), 0.0)
+
+    def test_nni_dm3(self):
+        actual_TreeNode = nni(self.pre3_nni_TreeNode, self.dm3, inplace=False)
+        self.assertAlmostEqual(actual_TreeNode.compare_tip_distances(
+            self.post3_nni_TreeNode), 0.0)
+        
+    def test_nni_trivial(self):
+        # No swaps are performed, but edge lengths are assigned.
+        data = [[0, 3, 2],
+                [3, 0, 3],
+                [2, 3, 0]]
+        dm = DistanceMatrix(data, list('abc'))
+        pre_str = "((c,b))a;"
+        pre_TreeNode = TreeNode.read(
+                io.StringIO(pre_str))
+        expected_str = "((c:1.0,b:2.0):1.0)a;"
+        expected_TreeNode = TreeNode.read(
+                io.StringIO(expected_str))
+        self.assertEqual(str(nni(pre_TreeNode, dm, inplace=False)),
+                         str(expected_TreeNode))
+
+    def test_nni_binary_flag(self):
+        # Flag prints out 'Tree needs to be a binary tree' and
+        # returns a None object.
+        data = [[0, 3],
+                [3, 0]]
+        dm = DistanceMatrix(data, list('ab'))
+        pre_str = "((b))a;"
+        pre_TreeNode = TreeNode.read(
+                io.StringIO(pre_str))
+        self.assertEqual(nni(pre_TreeNode, dm), None)
+
+    def test_nni_leaf_root_flag(self):
```

Comment:
The context manager approach for testing errors is more appropriate here as well.